Added opacity to cupertino switch when disabled

Robiquet opened the pull request:
## Description

Wrapped the widget in an opacity, if onChange is null the opacity is 0.4 otherwise it is 1.0

## Related Issues

https://github.com/flutter/flutter/issues/29164

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

Robiquet committed:
Changed author

xster commented:
Seems like there are 2 users involved in this. Make sure your GitHub account's email and your .gitconfig's [user] is the same user.

Robiquet commented:
@xster Am I right in thinking I can view that by running this:
` git config --list` ?
I can't any emails after running that command so I'm guessing I need to add it by doing something like this:
`git config --global user.email "MY_NAME@example.com"`
After doing that will I need to raise a new PR?

DanTup commented:
The author field on the commit seems to end with "MacBook-Pro.local".  I guess it made that up based on your username/machine name because you hadn't configured any. The command you gave above will fix this for future commits, but you'll need to fix this commit, which I think you can do with:

```
git commit --amend --author="Author Name <email@example.org>"
```

Then you'd have to force push (be careful doing this, if done in the wrong place or from the wrong branch, you could lose work).

Robiquet commented:
These tests were passing before I change the author on the commit, is it because the build went down? If so how do I get the tests to run again?

DanTup commented:
The failed tests looked like infrastructure issues (io errors/failed to launch instances) so I've restarted them.

Robiquet commented:
@DanTup Thanks, for future reference how did you do that?

DanTup commented:
@Robiquet it requires committer access to this repo (there is a re-run button on the Cirrus pages). Another way to do it would be to just push an empty commit commit (for ex. `git commit --allow-empty -m "Force build" && git push`).

xster commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -131,12 +131,15 @@ class CupertinoSwitch extends StatefulWidget {
-      vsync: this,
-      dragStartBehavior: widget.dragStartBehavior,
+    return Opacity(
+      opacity: widget.onChanged == null ? 0.4 : 1.0,
```
Document how you got to this value (it could just be eyeballing). This helps our next maintainer figure out how confident he/she needs to be before changing this.

xster commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -414,5 +414,27 @@ void main() {
     expect(value, isFalse);
   });
+  testWidgets('Switch is opaque when disabled', (WidgetTester tester) async {
+    final Key switchKey = UniqueKey();
```
You don't need this key if you're not using it

Robiquet replied:
Would you prefer if I did something like this?:
`          child: CupertinoSwitch(
            key: UniqueKey(),
            value: false,
            dragStartBehavior: DragStartBehavior.down,
            onChanged: null,
          ),`
If I remove the key parameter completely it creates a warning

xster replied:
A key parameter shouldn't be needed. What warning did it show?

Robiquet replied:
'Prefer const with constant constructors.'
Seems to be related to having 'onChanged' set to null. With this code there is a warning:
`
CupertinoSwitch(
            value: false,
            dragStartBehavior: DragStartBehavior.down,
            onChanged: null,
          ),
`
With this code there is not:
`
CupertinoSwitch(
            value: false,
            dragStartBehavior: DragStartBehavior.down,
            onChanged: (bool newValue) {},
          ),
`

xster replied:
oh, it just means you can instantiate the CupertinoSwitch (and likely its parent Center) via a compile time constant. (e.g. add a const in front of the Center constructor).

xster replied:
I made a change on your branch. The const should be at the highest node of the widget tree declaration possible.

xster commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -414,5 +414,27 @@ void main() {
+    await tester.pumpWidget(
+      Directionality(
+        textDirection: TextDirection.ltr,
+        child: StatefulBuilder(
```
You don't need this stateful builder if you're not using it

xster commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -414,5 +414,27 @@ void main() {
+      ),
+    );
+
+    expect(tester.widget<Opacity>(find.byType(Opacity).first).opacity, 0.4);
```
use .single instead of .first. This way, if in the future, another maintainer accidentally adds something into the widget tree that introduces another Opacity, this test would fail instead of succeeding but actually testing something else.

Robiquet replied:
Really appreciate your feedback on this PR. 

I can't figure this one out, I thought some like this would work:
`
expect(tester.widget<Opacity>(find.byType(Opacity).evaluate().single).opacity, 1.0);
`
However, the function is expecting something of type 'finder' not of type 'element'.

xster replied:
Oops my bad, Finder didn't actually implement single. It'd have to be 2 lines then:

```dart
expect(find.byType(Opacity), findsOneWidget);
expect(tester.widget<Opacity>(find.byType(Opacity).first).opacity, 1);
```

xster commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -414,5 +414,27 @@ void main() {
 
     expect(value, isFalse);
   });
+  testWidgets('Switch is opaque when disabled', (WidgetTester tester) async {
```
This description is reversed.

xster commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -414,5 +414,27 @@ void main() {
     expect(value, isFalse);
   });
+  testWidgets('Switch is opaque when disabled', (WidgetTester tester) async {
+    final Key switchKey = UniqueKey();
```
Please add another test to check that the switch is not translucent when enabled.

Robiquet replied:
Would you prefer if I did something like this?:
`
          child: CupertinoSwitch(
            key: UniqueKey(),
            value: false,
            dragStartBehavior: DragStartBehavior.down,
            onChanged: null,
          ),
`
If I remove the key completely it creates a warning

xster reviewed (approved):
Thanks for adding the test. This looks good to me. I just left some minor comments.

xster commented:
I think you resolved the comments but forgot to push your newest commits :)

Robiquet committed:
Modifications based on code review

Robiquet committed:
Fixed Cupertino switch warning

xster committed:
move the const

Robiquet committed:
Removed whitespace at end of line

xster commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -73,7 +73,9 @@ class CupertinoSwitch extends StatefulWidget {
   ///
-  /// If null, the switch will be displayed as disabled.
+  /// If null, the switch will be displayed as disabled with an opacity of
+  /// 0.5. This is opacity value appears to be accurate in comparison to the
```
Can you put the 0.5 as a constant in this file? You can then refer to it in this doc as `If null, the switch will be displayed as disabled with an opacity of [kCupertinoSwitchDisabledOpacity]`. 

We don't need to put the second sentence in the field doc. It can just be a code comment on the constant itself.

xster commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -131,12 +133,15 @@ class CupertinoSwitch extends StatefulWidget {
-      vsync: this,
-      dragStartBehavior: widget.dragStartBehavior,
+    return Opacity(
+      opacity: widget.onChanged == null ? 0.5 : 1.0,
```
Then you can use the constant here.

xster commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -414,5 +414,39 @@ void main() {
 
     expect(value, isFalse);
   });
+  testWidgets('Switch is translucent when disabled', (WidgetTester tester) async {
```
Small detail, leave a line between each test (to be consistent with the rest of the file).

xster commented on packages/flutter/test/cupertino/switch_test.dart:
```diff
@@ -414,5 +414,39 @@ void main() {
+    expect(find.byType(Opacity), findsOneWidget);
+    expect(tester.widget<Opacity>(find.byType(Opacity).first).opacity, 1.0);
+  });
 
```
Thanks for the tests. Looks good. Can you add one more test to check that going from enable to disabled and going from disabled to enabled take effect as you change?

Robiquet replied:
I'm having trouble achieving this, currently I'm trying to do something like this:

`
  testWidgets('Switch becomes opaque once enabled', (WidgetTester tester) async {
    bool isEnabled = false;

    await tester.pumpWidget(
      Directionality(
        textDirection: TextDirection.rtl,
        child: StatefulBuilder(
          builder: (BuildContext context, StateSetter setState) {
            return Center(
              child: CupertinoSwitch(
                value: false,
                dragStartBehavior: DragStartBehavior.down,
                onChanged: isEnabled ? (bool newValue) {} : null,
              ),
            );
          },
        ),
      ),      
    );

    expect(find.byType(Opacity), findsOneWidget);
    expect(tester.widget<Opacity>(find.byType(Opacity).first).opacity, 0.5);

    isEnabled = true;
    await tester.pump();

    expect(find.byType(Opacity), findsOneWidget);
    expect(tester.widget<Opacity>(find.byType(Opacity).first).opacity, 1.0);
  });
`
It's not working because the widget is not rebuilding with the new value. I can't seem to find any examples of this being done in the other test files. Do you know how I can do this or am I taking the wrong approach?

matthew-carroll replied:
To get the effect you desire, you should just be able to declare the same tree again, within another call to `tester.pumpWidget()`, but in the 2nd tree you should change the `onChanged` property to be non-null. Does that make sense?

Robiquet replied:
Yes I saw that in some other test files but I didn't think i would create the effect I needed. But after further thought it makes sense, thanks for you help

xster reviewed (approved):
Looks good. Just a few more comments.

Robiquet committed:
Switch constant and test spacing

Robiquet committed:
Removed whitespace

Robiquet committed:
Removed other space

Robiquet committed:
Whitespace

matthew-carroll commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -189,6 +192,7 @@ const double _kTrackInnerEnd = _kTrackWidth - _kTrackInnerStart;
 const double _kTrackInnerLength = _kTrackInnerEnd - _kTrackInnerStart;
 const double _kSwitchWidth = 59.0;
 const double _kSwitchHeight = 39.0;
+const double _kCupertinoSwitchDisabledOpacity = 0.5;//value appears to be accurate in comparison to the actual iOS switch
```
This comment is too long for the line. Would you mind moving the comment up above the constant declaration line and form it as a sentence? Also, would you mind pointing out that you eye-balled the value rather than used any particular tool? E.g.,

```dart
// Opacity of a disabled switch, as eye-balled from iOS.
const double _k....
```

Robiquet replied:
Made the change

matthew-carroll commented on packages/flutter/lib/src/cupertino/switch.dart:
```diff
@@ -73,7 +73,7 @@ class CupertinoSwitch extends StatefulWidget {
   /// value.
   ///
-  /// If null, the switch will be displayed as disabled.
+  /// If null, the switch will be displayed as disabled with an opacity of [kCupertinoSwitchDisabledOpacity]
```
The reference to the constant here won't work because the constant is private. Please consider a more ambiguous wording that allows us to avoid referencing the constant, e.g., "If null, the switch will be displayed as disabled, which has a reduced opacity."

Robiquet replied:
Made the change

Robiquet committed:
New tests and changes to docs

Tests for 'Switch turns translucent after becoming disabled' and 'Switch turns opaque after becoming enabled.' Also some small modifications to docs

Robiquet committed:
Whitespace removal
